### PR TITLE
First draft on the dedicated JSON page

### PR DESCRIPTION
**Pull request by emmanuelkeller**
Relate to #2043 (but don't close it)

### CODE REVIEW

File: docs/Indexing_JSON.md [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+FT.SEARCH myIdx '@name:(John)' RETURN 3 $.user.hp AS hitpoints
+1) (integer) 1
+2) "myDoc"
+3) 1) "hitpoints"
+   2) "1000"
+```
+
+### Highlighting
+
+On JSON, we can highlight any attribute as soon as he is indexed using the TEXT type.
```
Comment (oshadmi) [inline docs/Indexing_JSON.md:119]:
```suggestion
On JSON, we can highlight any attribute as soon as it is indexed using the TEXT type.
```

File: docs/Indexing_JSON.md [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+    "tag": ["foo","bar"],
+    "hp": "1000",
+    "dmg": "150"
+  }
+}
+```
+
+We can use `JSON.SET` to store in our database:
+
+    JSON.SET myDoc . '{"user":{"name":"John Smith","tag":["foo","bar"],"hp":1000, "dmg":150}}'
```
Comment (oshadmi) [inline docs/Indexing_JSON.md:56]:
```suggestion
    JSON.SET myDoc $ '{"user":{"name":"John Smith","tag":["foo","bar"],"hp":1000, "dmg":150}}'
```

File: docs/Commands.md [hunk 3]
```diff
@@ -169,17 +184,17 @@ FT.CREATE books-idx ON HASH PREFIX 1 book:details FILTER SCHEMA title TEXT categ
 
     * **SORTABLE**
 
-        Numeric, tag or text fields can have the optional SORTABLE argument that allows the user to later [sort the results by the value of this field](Sorting.md) (this adds memory overhead so do not declare it on large text fields).
+        Numeric, tag or text attributes can have the optional SORTABLE argument that allows the user to later [sort the results by the value of this field](Sorting.md) (this adds memory overhead so do not declare it on large text fields).
```
Comment (oshadmi) [inline docs/Commands.md:187]: GEO is also SORTABLE after [PR 2033](https://github.com/RediSearch/RediSearch/pull/2033)

File: docs/Commands.md [hunk 4]
```diff
@@ -75,11 +81,20 @@ Indexing books that have a "categories" field in which semicolons separate the v
...
+```sql
+FT.CREATE idx ON JSON SCHEMA $.title AS title TEXT $.categories AS categories TAG SORTABLE
+```
+
 #### Parameters
 
 * **index**: the index name to create. If it exists the old spec will be overwritten
 
-* **ON {structure}** currently supports only HASH (default)
+* **ON {structure}** currently supports HASH (default) and JSON.
```
Comment (oshadmi) [inline docs/Commands.md:94]: {structure} or {data type}?

File: docs/Commands.md [hunk 5]
```diff
@@ -75,11 +81,20 @@ Indexing books that have a "categories" field in which semicolons separate the v
 FT.CREATE books-idx ON HASH PREFIX 1 book:details FILTER SCHEMA title TEXT categories TAG SEPARATOR ";"
 ```
 
+Indexing a JSON document using JSON Path:
+
+```sql
+FT.CREATE idx ON JSON SCHEMA $.title AS title TEXT $.categories AS categories TAG SORTABLE
```
Comment (oshadmi) [inline docs/Commands.md:87]: I suggest to make sure this is auto tested exactly as documented (with search/aggregate command)
(even if a similar test exists, e.g., test_json.py:testDemo)

**Review by oshadmi — COMMENTED** [review]
👍🏼 Few comments

File: docs/Commands.md [hunk 6]
```diff
@@ -940,14 +981,14 @@ OK or an error.
...
-FT.TAGVALS {index} {field_name}
+FT.TAGVALS {index} {attribute_name}
 ```
 
 #### Description
 
 Returns the distinct tags indexed in a [Tag field](Tags.md).
 
-This is useful if your tag field indexes things like cities, categories, etc.
+This is useful if your tag attribute indexes things like cities, categories, etc.
```
Comment (banker) [inline docs/Commands.md:991]:
```suggestion
This is useful if your tag indexes things like cities, categories, etc.
```

Comment (banker) [reply docs/Commands.md:991]: @emmanuelkeller Do you agree with this piece of documentation?

File: docs/Commands.md [hunk 7]
```diff
@@ -940,14 +981,14 @@ OK or an error.
 #### Format
 
 ```
-FT.TAGVALS {index} {field_name}
+FT.TAGVALS {index} {attribute_name}
 ```
 
 #### Description
 
 Returns the distinct tags indexed in a [Tag field](Tags.md).
```
Comment (banker) [inline docs/Commands.md:989]:
```suggestion
Returns the distinct set of values indexed in a [Tag field](Tags.md).
```

File: docs/Commands.md [hunk 8]
```diff
@@ -970,7 +1011,7 @@ Array Reply: All the distinct tags in the tag index.
 
 #### Complexity
 
-O(n), n being the cardinality of the tag field.
+O(n), n being the cardinality of the tag attribute.
```
Comment (banker) [inline docs/Commands.md:1014]: @emmanuelkeller This doesn't make sense to me. How can a field have a cardinality? Is `n` the number of distinct values stored? Why not say it more directly?

File: docs/Indexing_JSON.md [hunk 9]
```diff
@@ -0,0 +1,142 @@
+# Indexing JSON documents
+
+As a secondary index, RediSearch aims to index data structures available in Redis.
```
Comment (banker) [inline docs/Indexing_JSON.md:3]:
```suggestion
In addition to indexing Redis hashes, RediSearch also indexes JSON. To index JSON, you must use the RedisJSON module.
```

File: docs/Indexing_JSON.md [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+In its first implementations (1.x, 2.0) RediSearch supported only hashes.
```
Comment (banker) [inline docs/Indexing_JSON.md:4]:
```suggestion

```

File: docs/Indexing_JSON.md [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+Associated with RediJSON (version 2.0 or more recent), RediSearch is able to index JSON documents.
```
Comment (banker) [inline docs/Indexing_JSON.md:5]:
```suggestion

```

File: docs/Indexing_JSON.md [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+
+## Prerequisite
+
+What do you need to start indexing JSON documents ?
```
Comment (banker) [inline docs/Indexing_JSON.md:9]:
```suggestion
What do you need to start indexing JSON documents?
```

File: docs/Indexing_JSON.md [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+
+- Redis 6.0x (or more recent)
```
Comment (banker) [inline docs/Indexing_JSON.md:11]:
```suggestion
- Redis 6.x or later
```

File: docs/Indexing_JSON.md [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+- The module RediSearch 2.2 (or more recent)
```
Comment (banker) [inline docs/Indexing_JSON.md:12]:
```suggestion
- RediSearch 2.2 or later
```

File: docs/Indexing_JSON.md [hunk 15]
```diff
@@ -0,0 +1,142 @@
...
+- The module RediJSON 2.0 (or more recent)
```
Comment (banker) [inline docs/Indexing_JSON.md:13]:
```suggestion
- RediJSON 2.0 or later
```

File: docs/Indexing_JSON.md [hunk 16]
```diff
@@ -0,0 +1,142 @@
...
+
+## Creating an index for JSON documents indexing
```
Comment (banker) [inline docs/Indexing_JSON.md:15]:
```suggestion
## How to index JSON documents
```

**Review by banker — CHANGES_REQUESTED** [review]
Initial review @emmanuelkeller. I'll finish the rest in the morning.

**Review by banker — CHANGES_REQUESTED** [review]

File: docs/Commands.md [hunk 17]
```diff
@@ -69,24 +75,33 @@ Indexing only books that have a subtitle:
...
+!!! info "ON JSON"
+    Require the [RedisJSON](https://redisjson.io) module to be installed.
 
 * **PREFIX {count} {prefix}** tells the index which keys it should index. You can add several prefixes to index. Since the argument is optional, the default is * (all keys)
 
 * **FILTER {filter}** is a filter expression with the full RediSearch aggregation expression language. It is possible to use @__key to access the key that was just added/changed. A field can be used to set field name by passing `'FILTER @indexName=="myindexname"'`
 
 * **LANGUAGE {default_lang}**: If set indicates the default language for documents in the index. Default to English.
-* **LANGUAGE_FIELD {lang_field}**: If set indicates the document field that should be used as the document language.
+* **LANGUAGE_FIELD {lang_attribute}**: If set indicates the document attribute that should be used as the document language.
```
[the remark this answers is not in the document]
Comment (emmanuelkeller) [reply docs/Commands.md:104]: > @emmanuelkeller Can we keep as field? We're not changing the name of the options (LANGUAGE_FIELD), so why use another word in the definition?

@banker We are going to change the keywords in a upcoming milestone. It is part of the spec, but not yet available in this one. Both LANGUAGE_FIELD and LANGUAGE_ATTRIBUTE will be usable

File: docs/Commands.md [hunk 18]
```diff
@@ -216,35 +240,41 @@ OK or an error
...
+If an attribute fails to be indexed (for example, if a numeric attributes gets a string value) the whole document is not indexed. `FT.INFO` provides the number of document-indexing-failures under `hash_indexing_failures`.
 
-If `LANGUAGE_FIELD`, `SCORE_FIELD`, or `PAYLOAD_FIELD` were used with `FT.CREATE`, the document will extract the properties. A field can be used to get the name of the index it belongs to.
+If `LANGUAGE_FIELD`, `SCORE_FIELD`, or `PAYLOAD_FIELD` were used with `FT.CREATE`, the document will extract the properties. A attribute can be used to get the name of the index it belongs to.
 
 !!! warning "Schema mismatch"
-    If a value in a hash does not match the schema type for that field, indexing of the hash will fail. The number of 'failed' document is under `hash_indexing_failures` at `FT.INFO`.
+    If a value in a hash does not match the schema type for that attribute, indexing of the hash will fail. The number of 'failed' document is under `hash_indexing_failures` at `FT.INFO`.
```
[the remark this answers is not in the document]
Comment (emmanuelkeller) [reply docs/Commands.md:267]: > Isn't this also true for a JSON document, @emmanuelkeller ?
@banker Yes it is. The content of FT.INFO will be updated in the upcoming milestone

File: docs/Commands.md (hunk 4 above)
Comment (emmanuelkeller) [reply docs/Commands.md:94]: @oshadmi Good point I think we can replace structure by datatype.
@banker  Yes, HASH is still a default. ON not required

**Pull request closed without merging**